### ISSUE TITLE
github.com/Azure/go-autorest/autorest/adal@v0.9.18 has been retracted

### ISSUE THREAD

**Issue by apparentlymart** [opentofu/opentofu#2187]
Our `go.mod` currently records that we indirectly depend on v0.9.18 of `github.com/Azure/go-autorest/autorest/adal`.

This version has apparently been retracted:

```
go: warning: github.com/Azure/go-autorest/autorest/adal@v0.9.18: retracted by module author: retracted due to token refresh errors
go: to switch to the latest unretracted version, run:
        go get github.com/Azure/go-autorest/autorest/adal@latest
```

Running the suggested command upgrades to v0.9.24, but since this dependency seems to exist only to serve the `azurerm` backend I'm not equipped to test that upgrade myself. (I don't have an Azure account to test with.)

This is only a warning, so we are not blocked by this retraction but seems worth investigating nonetheless.

**Comment by cam72cam** [comment]
It looks like that whole module is deprecated:
go: module github.com/Azure/go-autorest/autorest/adal is deprecated: use github.com/Azure/azure-sdk-for-go/sdk/azidentity instead.

**Comment by cam72cam** [comment]
https://azure.microsoft.com/en-us/updates?id=support-for-azure-sdk-libraries-that-do-not-conform-to-our-current-azure-sdk-guidelines-will-be-retired-as-of-31-march-2023

**Comment by lbordowitz** [comment]
This is tangentially related to #2769 : there are a number of unused variables in the Azure backend from clients that aren't attached to the ArmClient structure. However, the mixture of different clients seems unintentional; compared to other remote backends, I'd say the complexity of the different clients constitutes tech debt.

In fact, I have half a mind to re-write the Azure backend using github.com/Azure/azure-sdk-for-go/. Main hurdles would be auth option parity and ensuring a sufficient amount of testing is done.

**Comment by cam72cam** [comment]
We've had one person [start to update](https://github.com/opentofu/opentofu/pull/2512) some of the libraries for new functionality, but it's still tied to HC's azure libs.  IMO that's both a blessing and a curse :)

Unfortunately that PR has stalled out.  We also don't have an abundance of Azure experience on the maintainer team right now.

I'd be open to a pure sdk rewrite of this backend, with the caveat that it needs to maintain bug-for-bug compatibility where possible.  It would be a very high bar for testing, but I'd be willing to advocate for it if you think you've got the time / energy /  experience to take this on.

**Comment by cam72cam** [comment]
As discussed in our Weekly Dev Sync, @lbordowitz will give his proposal a try with support from the Maintainers.  It's a significant change, but one that would be well appreciated.

**Comment by lbordowitz** [comment]
The migration to a new client (azure-sdk-for-go) is a work-in-progress. Part of this involves updating the Service Version. Various APIs were used prior to the upgrade; I've seen 2018-11-09, also 2017-03-09 and 2021-01-01, and the present service version in the client is 2025-05-05. Somewhere between these versions (I'm guessing 2024-11-04) an auth incompatibility emerged.

When I test the backend, I get the following error:

```
│ Error: Failed to get existing workspaces: GET https://*****.blob.core.windows.net/tfstate
│ --------------------------------------------------------------------------------
│ RESPONSE 403: 403 This request is not authorized to perform this operation using this permission.
│ ERROR CODE: AuthorizationPermissionMismatch
```

This was strange because:

a. I was the owner of this subscription/tenant 
b. It worked before on the azure backend, prior to my changes.

 [This error has been encountered before](https://learn.microsoft.com/en-us/answers/questions/2087624/authorizationpermissionmismatch-while-uploading-fi) when someone upgraded their version. Another hint comes from the [documentation](https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-v10?tabs=dnf#run-azcopy):

> As an owner of your Azure Storage account, you aren't automatically assigned permissions to access data. Before you can do anything meaningful with AzCopy, you need to decide how you'll provide authorization credentials to the storage service.

Upon providing my owner account the "Storage Blob Data Contributor" on the tfstate container, I was able to initialize the backend without any further code changes.

I'm worried that this could mean, in order to make this change to the azure backend, we will have to either

- Provide this as a separate backend
- Somehow make the background API version-compatible with the previous one

Both options are unpleasant. I would like a second opinion on this. I'm currently developing this feature on [a personal branch](https://github.com/lbordowitz/opentofu/tree/azure_refactor), under the `assure` backend name. My strategy is to have something that barely works that I can iterate on, hardening it through testing. You should be able to compile it presently and initialize the backend; it's currently failing on finding the lock id in the metadata for a different reason: see [here](https://github.com/Azure/azure-sdk-for-go/issues/24517), [here](https://github.com/Azure/azure-sdk-for-go/issues/16791) and [here](https://github.com/Azure/azure-sdk-for-go/issues/645) for similar fixed issues; I might just change the lock_id key to use [go's CanonicalMIMEHeader case](https://pkg.go.dev/net/textproto#CanonicalMIMEHeaderKey).

**Comment by lbordowitz** [comment]
The solution to the problem in the previous comment: the authentication providers are actually a bootstrap to obtain the Storage Account's Shared Keys for Shared Key Authentication. We can do this kind of authentication, too, and it results in the same backend authentication set up. So that's now a non-issue.

To the other problem about setting metadata: The header MUST be the same case. This change requires a change in the azure blob client. This is a generated client, so a change needs to be made in the autorest.go generator in order for this to be fixed on the azure blob client. I'm working with the maintainers there on an [issue](https://github.com/Azure/autorest.go/issues/1600) and a [pull request](https://github.com/Azure/autorest.go/pull/1601). I still need to do a bit more testing to check that this is the correct fix, but I'm feeling fairly confident.

After this is done, the plan is to continue implementing tests, fixing problems, and introducing more error-handling paths. After that, I'd like to implement the rest of the authentication methods.

What would be the best venue to find people who use the `azurerm` backend? If I could get a small "alpha testing" group with the different auth methods, it would help bolster the change's correctness.

**Comment by ImIOImI** [comment]
In preparation for this I picked up where I left off with my effort to make several examples of azure bootstraps with different login methods and I'm working through terratest to automate testing. I'm going to be out of country (in Belize) for the next 2 weeks, so don't expect to hear anything from me soon.

The plan is to test a couple different login methods via github workflows. Fortunately, all of the most relevant ones are already covered in the [GitHub azure login action](https://github.com/marketplace/actions/azure-login) (which I make full use of in my company's workflows). 

In this action you can see examples of:
1. [Connecting via OpenID](https://github.com/marketplace/actions/azure-login#login-with-openid-connect-oidc-recommended) If you're going to run tofu on GitHub hosted runners, this is the way you should do it. This is not particularly hard (imo) to setup and there seems to be lots of documentation on it. This is prohibitive for many data plane operations however, that need network access (like writing to a storage account providing you don't have said account publically accessible). 
2. [Logging in with a service principal secret](https://github.com/marketplace/actions/azure-login#login-with-a-service-principal-secret) This is the easiest to setup, so I suspect it's fairly common despite the fact that saving secrets kinda suck. I also suspect this will be the first way you should test changes.
3. [Logging in with a system assigned user identity](https://github.com/marketplace/actions/azure-login#login-with-system-assigned-managed-identity) This kind of identity is assigned by Azure. Its for things like authorizing all the nodes in a cluster to have permission to do things. I regard this, generally, as anti-pattern for tofu because giving a cluster the ability to run any infra operation in your tennant/subscription seems incredibly dangerous to me. 
4. [Logging in with a user managed identity](https://github.com/marketplace/actions/azure-login#login-with-user-assigned-managed-identity) This is what I primarily use, and it's really easy to manage on k8s runners so I expect it to be a fairly common use case. This will be harder to test though, since you will need to be running this on an Azure hosted machine (I believe). The way I employ it is I give kubernetes service accounts the ability to assume a tofu infra role (as opposed to an entire machine like in method 3) and that authorizes that service account to run the tofu workload at an elevated access level. 

Methods 3 and 4 are going to require actual machines in Azure to login to test so that will add a level of complication. I'm happy to build out some tofu examples to help with this and/or test builds at work... but it will be a while before I can help on that effort.
